src/common/meson.build

```diff
@@ -0,0 +1,384 @@
+        endif
+    endforeach
+
+    if have_byteswap_h
+        bswap_fcts = [
+            'bswap_16',
+            'bswap_32',
+            'bswap_64'
+        ]
+        foreach f : bswap_fcts
+            bswap_src = '''
+#include <byteswap.h>
+int main()
+{
+	@0@(42);
+	return 0;
+}
+'''
+            if cc.links(bswap_src.format(f), name: f)
+                config_h.set10('HAVE_' + f.to_upper(), true,
+                    description: 'Define to 1 if ' + f + '() is available.'
+                )
+            endif
+        endforeach
+    endif
+endif
+
+fast_unaligned_access = false
+if get_option('unaligned-access').auto()
+    unaligned_access_cpus = [
+        'ppc',
+        'ppc64',
+        'x86',
+        'x86_64',
+    ]
+    if host_machine.cpu_family() in unaligned_access_cpus
+        config_h.set10('TUKLIB_FAST_UNALIGNED_ACCESS', true,
+            description: 'Define to 1 if the system supports fast unaligned access to 16-bit, 32-bit, and 64-bit integers.'
+        )
+        fast_unaligned_access = true
+    endif
+
+    if not fast_unaligned_access
+        unaligned_access_cpus = [
+            'aarch64',
+            'arm',
+            'riscv32',
+            'riscv64',
+        ]
+        if host_machine.cpu_family() in unaligned_access_cpus
+            ua_src = '''
+#if !defined(__ARM_FEATURE_UNALIGNED) \
+		&& !defined(__riscv_misaligned_fast) \
+		&& !defined(_MSC_VER)
+compile error
+#endif
+int main(void) { return 0; }
+'''
+            if cc.compiles(ua_src, name: 'fast unaligned access for ' + host_machine.cpu_family())
+                config_h.set10('TUKLIB_FAST_UNALIGNED_ACCESS', true,
+                description: 'Define to 1 if the system supports fast unaligned access to 16-bit, 32-bit, and 64-bit integers.'
+                )
+                fast_unaligned_access = true
+            endif
+        endif
+    endif
+elif get_option('unaligned-access').enabled()
+    config_h.set10('TUKLIB_FAST_UNALIGNED_ACCESS', true,
+        description: 'Define to 1 if the system supports fast unaligned access to 16-bit, 32-bit, and 64-bit integers.'
+    )
+    fast_unaligned_access = true
+endif
+
+if get_option('unsafe-type-punning')
+    config_h.set10('HAVE___BUILTIN_ASSUME_ALIGNED', true,
+        description: 'Define to 1 if the GNU C extension __builtin_assume_aligned is supported.'
+                    )
+endif
+
+# tuklib_mbstr.m4
+
+### FIXME : mbrtowc is supported in C99 spec and on Windows
+if cc.has_header('wchar.h')
+    config_h.set10('HAVE_WCHAR_H', true,
+        description: 'Define to 1 if the header file <wchar.h> is available.'
+    )
+
+    if cc.has_type('mbstate_t', prefix : '#include <wchar.h>') and cc.has_function('mbrtowc', prefix : '#include <wchar.h>')
+        config_h.set10('HAVE_MBRTOWC', true,
+            description: 'Define to 1 if the mbrtowc() function is available.'
+        )
+    endif
+
+    if cc.has_function('wcwidth',
+        args : '-D_XOPEN_SOURCE',
```

Comment:
It's incorrect to add `-D_XOPEN_SOURCE` into a single test. The actual C code won't use that `#define`, and thus the test and the C code wouldn't be in sync. If you added this because the wcwidth(3) man page in the man-pages package lists it, note that the same line on the man page tells to look at feature_test_macros(7) which tells that `_GNU_SOURCE` already covers this.

The codebase relies on some platform-specific extensions when they are available, thus it's not practical to build it with something like `-D_XOPEN_SOURCE=600`. Instead, all system-specific extensions are hoped to be available which means using, for example, `#define _GNU_SOURCE 1`. Also, some of the now-POSIX features were available on some systems before they were standardized in POSIX, so this approach works better on such systems.

Thus, it's essential that `_GNU_SOURCE` and such things are added to project-wide defines before running any checks for available functions or other features, and the checks then don't use any function-specific extra args unless the C code uses such extras too.

Reply:
I, of course, agree with you on all of the above - in addition to it being cleaner to be consistent, but it's _also_ an issue for C++ with ODR too. And sometimes projects have to be built as C++ even if they're written in C.